### PR TITLE
Nanosecond resolution for C++

### PR DESCRIPTION
**Pull request by javmorcas**
This pull request is coming from the issue #253.

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -151,6 +155,10 @@ function usage()
     echo "                                 binding. Default 10MB                          "
     echo "    --osx-shmem-shmmax <size>    Maximum segment size for shared memory in OSX  "
     echo "                                 in bytes. Default 400MB                        "
+    echo "    --ns-resolution              Try to use the system real-time clock to get   "
+    echo "                                 nano-second resolution.                        "
+    echo "                                 For the Classic C++ Implementation only.       "
```
Comment (alfonso-soto) [inline build.sh]: You said "Classic C++ Implementation only" but I also see it enabled on the CPP moderm.

File: srcCpp/Infrastructure_micro.cxx [hunk 2]
```diff
@@ -88,6 +89,10 @@ unsigned long long PerftestClock::getTimeUsec()
     QueryPerformanceCounter(&ticks);
     return ticks.QuadPart / (unsigned long long) (_frequency /1000000.0);
 
+    #endif
```
Comment (alfonso-soto) [inline srcCpp/Infrastructure_micro.cxx]: As a suggestion, for more readable code
`#endif  /* RTI_WIN32 */`

File: srcCpp/Infrastructure_micro.cxx [hunk 3]
```diff
@@ -88,6 +89,10 @@ unsigned long long PerftestClock::getTimeUsec()
...
+  #else
+    clock_gettime(CLOCK_MONOTONIC, &timeStruct);
+    return (timeStruct.tv_sec * ONE_BILLION) + timeStruct.tv_nsec;
   #endif
```
Comment (alfonso-soto) [inline srcCpp/Infrastructure_micro.cxx]: `#endif /* RTI_PERFTEST_NANO_CLOCK */`

File: srcCpp/Infrastructure_micro.h [hunk 4]
```diff
@@ -43,6 +43,13 @@ inline RTI_BOOL PerftestSemaphore_take(PerftestSemaphore *sem, int timeout)
 }
 
 #define PERFTEST_DISCOVERY_TIME_MSEC 1000 // 1 second
+#define ONE_BILLION  1000000000L // 1 billion (US)
```
Comment (alfonso-soto) [inline srcCpp/Infrastructure_micro.h]: // 1 billion (US) == 1 second

File: srcCpp/Infrastructure_pro.h [hunk 5]
```diff
@@ -66,23 +66,34 @@ inline bool PerftestSemaphore_give(PerftestSemaphore *sem)
         PerftestSemaphore_take(sem, PERFTEST_SEMAPHORE_TIMEOUT_INFINITE)
 
 #define PERFTEST_DISCOVERY_TIME_MSEC 1000 // 1 second
+#define ONE_BILLION  1000000000L // 1 billion (US)
+
+#ifndef RTI_PERFTEST_NANO_CLOCK
+  #define PERFT_TIME_UNIT "us"
+#else
+  #define PERFT_TIME_UNIT "ns"
+#endif
```
Comment (alfonso-soto) [inline srcCpp/Infrastructure_pro.h]: If this is the same that is defined on Infrastructure_micro.h we could move it into Infrastructure_common.h

File: srcCpp/Infrastructure_pro.cxx [hunk 6]
```diff
@@ -45,14 +53,19 @@ PerftestClock &PerftestClock::getInstance()
...
 {
+  #ifndef RTI_PERFTEST_NANO_CLOCK
     clock->getTime(clock, &clockTimeAux);
     RTINtpTime_unpackToMicrosec(
             clockSec,
             clockUsec,
             clockTimeAux);
     return clockUsec + 1000000 * clockSec;
+  #else
+    clock_gettime(CLOCK_MONOTONIC, &timeStruct);
```
Comment (alfonso-soto) [inline srcCpp/Infrastructure_pro.cxx]: Check for error:
```
RETURN VALUE
       clock_gettime(), clock_settime(), and clock_getres() return 0 for success, or -1 for failure (in which case errno is set appropriately).
```

File: srcCpp/perftest_publisher.cxx [hunk 7]
```diff
@@ -1367,18 +1367,25 @@ class LatencyListener : public IMessagingCB
...
+               " Latency: Ave %6.0lf " PERFT_TIME_UNIT
+               " Std %6.1lf " PERFT_TIME_UNIT
+               " Min %6lu " PERFT_TIME_UNIT
+               " Max %6lu " PERFT_TIME_UNIT
+               " 50%% %6lu " PERFT_TIME_UNIT
+               " 90%% %6lu " PERFT_TIME_UNIT
+               " 99%% %6lu " PERFT_TIME_UNIT
+               " 99.99%% %6lu " PERFT_TIME_UNIT
+               " 99.9999%% %6lu " PERFT_TIME_UNIT
+               " %s\n",
```
Comment (alfonso-soto) [inline srcCpp/perftest_publisher.cxx]: Much much cleaner!! I like it.

File: srcCpp/perftest_publisher.cxx [hunk 8]
```diff
@@ -1367,18 +1367,25 @@ class LatencyListener : public IMessagingCB
...
+        printf("Length: %5d"
+               " Latency: Ave %6.0lf " PERFT_TIME_UNIT
+               " Std %6.1lf " PERFT_TIME_UNIT
+               " Min %6lu " PERFT_TIME_UNIT
+               " Max %6lu " PERFT_TIME_UNIT
+               " 50%% %6lu " PERFT_TIME_UNIT
+               " 90%% %6lu " PERFT_TIME_UNIT
+               " 99%% %6lu " PERFT_TIME_UNIT
+               " 99.99%% %6lu " PERFT_TIME_UNIT
+               " 99.9999%% %6lu " PERFT_TIME_UNIT
```
Comment (alfonso-soto) [inline srcCpp/perftest_publisher.cxx]: Probably we need to go for %6llu?

The maximum value on an unsigned long is 4294967295 that is equal to 4.29 seconds.

File: build.sh (hunk 1 above)
Comment (javmorcas) [reply build.sh]: Where? It should not be

File: srcCpp/Infrastructure_micro.cxx (hunk 2 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_micro.cxx]: Done

File: srcCpp/Infrastructure_micro.cxx (hunk 3 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_micro.cxx]: Done

File: srcCpp/Infrastructure_micro.h (hunk 4 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_micro.h]: Done

File: srcCpp/Infrastructure_pro.cxx (hunk 6 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_pro.cxx]: I will leave this open, as I would not know what to do with that error and I am hesitant to have to do that extra assignment and checking.

File: srcCpp/Infrastructure_pro.h (hunk 5 above)
Comment (javmorcas) [reply srcCpp/Infrastructure_pro.h]: Agreed. Will do.

File: srcCpp/perftest_publisher.cxx (hunk 8 above)
Comment (javmorcas) [reply srcCpp/perftest_publisher.cxx]: Yes, most likely, I will add this as a different issue, as it involves refactoring the latency_ave, min, max, and history to be unsigned long longs

Comment (javmorcas) [reply srcCpp/perftest_publisher.cxx]: #256

**Pull request merged**